### PR TITLE
Extend ChannelType by command options

### PR DESCRIPTION
**Pull request by htreu**
This is WIP and open for discussion.

This addresses #5099 by adding command options as an alternative to a state description.
The implementation is naive and straigt forward.
Command options will be rendered as push buttons by UIs and send the corresponding command value as a command to the channel.
With this proposal, the state of the channel will not be represented in the UI, so ThingHandlers may not even update the state.

Signed-off-by: Henning Treu <henning.treu@telekom.de>

### CODE REVIEW

**Comment by htreu** [comment]
Paper UI will now render push-buttons for Switch & String items when the corresponding ChannelType defines commandOptions.

**Comment by htreu** [comment]
This is an example of the Sonos binding's "Clear Queue" channel:

screen shot 2018-02-22 at 12 03 16

**Comment by triller-telekom** [comment]
I just tested this PR and it works for me. But before merging I would say that the feature should be "complete" in terms of that BasicUI and ClassicUI should also render buttons for command options.

File: bundles/core/org.eclipse.smarthome.core.thing.xml/src/main/java/org/eclipse/smarthome/core/thing/xml/internal/ChannelTypeConverter.java [hunk 1]
```diff
@@ -142,6 +173,7 @@ protected ChannelTypeXmlResult unmarshalType(HierarchicalStreamReader reader, Un
         String description = super.readDescription(nodeIterator);
         String category = readCategory(nodeIterator);
         Set<String> tags = readTags(nodeIterator);
+        List<CommandOption> commandOptions = readCommandOptions(nodeIterator);
```
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core.thing.xml/src/main/java/org/eclipse/smarthome/core/thing/xml/internal/ChannelTypeConverter.java]: Even though we do not yet have anything besides the options - wouldn't it be more symmetrical to have a "commandDescription" next to the state and event descriptions?

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/types/CommandOption.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+    /**
+     * The command which will be send to the Channel
+     */
+    private final String command;
+
+    /**
+     * The name of the command which will be displayed in the UI.
+     */
+    private final String name;
```
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/types/CommandOption.java]: In StateOption, this is called "label".

File: extensions/binding/org.eclipse.smarthome.binding.sonos/ESH-INF/thing/channels.xml [hunk 3]
```diff
@@ -31,9 +31,12 @@
 	</channel-type>
 
 	<channel-type id="clearqueue" advanced="true">
-		<item-type>Switch</item-type>
+		<item-type>String</item-type>
```
Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.sonos/ESH-INF/thing/channels.xml]: This change can break existing setups.

File: bundles/test/org.eclipse.smarthome.magic/ESH-INF/thing/channel-types.xml [hunk 4]
```diff
@@ -51,13 +51,11 @@
 		<item-type>String</item-type>
 		<label>Alert</label>
 		<description>The alert channel allows a temporary change to the bulb’s state.</description>
-		<state>
```
Comment (kaikreuzer) [inline bundles/test/org.eclipse.smarthome.magic/ESH-INF/thing/channel-types.xml]: This breaks backward compatibility as it removes the state options, which are still a valid information.

File: extensions/binding/org.eclipse.smarthome.binding.sonos/ESH-INF/thing/channels.xml (hunk 3 above)
Comment (htreu) [reply extensions/binding/org.eclipse.smarthome.binding.sonos/ESH-INF/thing/channels.xml]: thanks, this was not intended is not even necessary. Switch items will also get the push buttons.

**Comment by htreu** [comment]
> Wrt rendering buttons: So far, we render drop-down boxes, which is also a valid solution - it only fails for "pure command channels" which do not have any state and thus a selection should not stay active. But this is going to be addressed with #5011.

A drop-down box really fails at being a command input. Selecting a value from the list is more of a "change value" notion then pushing a button which has a meaning of "do it now".

What might be a discriminating property for "pure command channels" is the fact that only one command is available. This would mean the `<command-options>` list should rather be a single `<command-option value="clear">Clear Queue</...>` entry.

**Comment by maggu2810** [comment]
> With this proposal, the state of the channel will not be represented in the UI, so ThingHandlers may not even update the state.

So it is exclusive? Using command options will hide the state in the UI?

**Comment by kaikreuzer** [comment]
> Using command options will hide the state in the UI?

It definitely must not. There are clearly channels where you want command options, but which also have a state.

> A drop-down box really fails at being a command input.

Buttons really fail once you have too many of them...

> What might be a discriminating property for "pure command channels" is the fact that only one command is available. 

Why this? Id rather think that "pure command channels" usually have many options as it does not make sense to create 3 command channels for a thing, if you can create a single one that can handle 3 different commands.

**Comment by htreu** [comment]
>> Using command options will hide the state in the UI?

>It definitely must not. There are clearly channels where you want command options, but which also have a state.

Then you are more close to select a new state from the options. Also: examples please.

>>A drop-down box really fails at being a command input.

>Buttons really fail once you have too many of them...

Here single-command-only-channels come in (see below). Besides: having too many buttons doesn't make drop downs any better ;-)

> Why this? Id rather think that "pure command channels" usually have many options as it does not make sense to create 3 command channels for a thing, if you can create a single one that can handle 3 different commands.

Why would you want 1 channel for arbitrary, unrelated commands? Having > 1 command on a channel may be an indicator for the channel to be a "normal" state channel.
I'd like to see some examples to see which channel category fits the different situations.

Example:
Sonos binding, command: "Clear Queue"
* This channel should not have other commands
* The channel does not change its own state

**Comment by kaikreuzer** [comment]
> Then you are more close to select a new state from the options. Also: examples please.

Almost all that you can currently find. E.g. "alert" channel from the hue binding.

> Here single-command-only-channels come in (see below). Besides: having too many buttons doesn't make drop downs any better ;-)

I don't understand what kind of widget you then suggest for e.g. 5 command options?
I think the [Scene Selection](http://demo.openhab.org:8080/basicui/app?w=0202&sitemap=demo) works pretty good here.

> Why would you want 1 channel for arbitrary, unrelated commands? 

Why would you want to have multiple channels, when all that they provide is a distinct state (which you do not have)?

See https://github.com/openhab/openhab2-addons/pull/2345#discussion_r164434998:
_Anyhow, it does not make sense to have two commands (ON and OFF) for every single trigger here. If we say that the state does not make any sense (as it gets out of sync too easily), we should actually not model any separate channels at all, but rather have a single "command channel", which would take Strings like PLAY/PAUSE/RESET (and the user could have a single button for each of those options in the UI)._

**Comment by htreu** [comment]
> E.g. "alert" channel from the hue binding

As I understand the hue `alert` channel its something which can be switched on/off and a UI should reflect the current state. I see it being modelled as a "normal" on/off channel linked to a Switch item (like it is right now).

> the Scene Selection works pretty good

No, ist does not: How do you select an already selected preset _again_? With push buttons you would simply press it once more. Also, selecting a preset does not mean a distinct state to change to the preset name. It would only be the status channel "last preset selected". A scene/preset selection changes state on a different channel.
Example: The tuning of an analog radio station would have the frequency channel and a command-options preset channel. Selecting on of the preset radio stations would change state on the frequency channel but leave no information on the preset channel. Here you _could_ mark a preset as selected when the frequency channel tunes into the corresponding frequency. I would consider this not to be necessary as the state is display elsewhere. And coming to the scene selection again you would not want to highlight a specific scene if by accident all corresponding states suddenly match.

The Lametric Time example is nice and another counter example to the single-command-only channel idea:
* The countdown app receives commands start/pause/reset
* There is no feedback channel, so no state which tells if the countdown is running or is in the reset state
* > 1 commands which belong together

I agree, a command channel with many options/buttons is not nice to render (right now) but I'm sure we will come up with a solution.

**Comment by davidgraeff** [comment]
This is exactly what I meant with #4272. Hopefully this gets merged some time :)

**Comment by mgeramb** [comment]
Hi all,
my binding [AmazonEchoControl](https://community.openhab.org/t/beta-openhab2-amazonechocontrol-binding-controlling-alexa-from-openhab2/37844) is also an example for the need of command channels. I have different command (start playing news flash, start playing weather information, start playing flash briefing). But I have no feedback from the device what is currently playing. Currently I use switch channels where the binding set the value automatically back to Off. But I think this is not a nice solution because the thing gets more and more channels. So I'am currently looking for a other solution. I think the command option and a single channel will help. 

But I'am not sure if the commands should be only static. I have other channels which get the available options from the device. E.g. playAlarmSound, the available sounds are provided with a DynamicStateDescriptionProvider, it would be nice if this feature could also be used for these channels in the future (Currently I have no need, because for these types of channels, I get a feedback from the device).

**Comment by Tony-Hu** [comment]
I'm wondering that if the stateless channel is an essential part of some certain scenarios. Sometimes, either users or developers may not care about the current state a channel holds. However, shall we provide a way to block the state in lower level functions?
>This is an example of the Sonos binding's "Clear Queue" channel:

Perhaps we can consider the ``clearqueue`` channel has a state to indicate how many songs in the play queue by using non-negative numbers. And some negative numbers to indicate the queue is being cleared or it's failed. Although it's rarely happened for the failed execution of clear queue, it will make users unhappy. 
For those users who don't about some certain states, we can provide an option to hide those in both web UIs and mobile clients.
And for the developers, they can just ignore the state in various ways.

**Comment by kaikreuzer** [comment]
@htreu Could you please rebase this branch?
Wdyt about the suggestion of a `DynamicCommandDescriptionProvider`?
Apart from this, is there anything else open after the discussion?

**Comment by htreu** [comment]
I did rebase the branch and adopt the latest Paper UI refactorings.

Open issues:

* Rename `commandOptions` to `commandDescription`
* Provide Thing DSL enhancement for textual configuration
* provide and evaluate `DynamicCommandDescriptionProvider`
* Implement Classic UI & Basic UI

**Comment by davidgraeff** [comment]
@htreu This would be nice to have as well for a lot of bindings ^^

**Comment by htreu** [comment]
Since I#m about to implement at least some of the missing pieces: What is you expectation on how to access the `CommandDescription`? Right now it is only accessible through the ChannelType. I suppose we want the information about commands and a localized description to be present on the item domain? wdyt? /cc @kaikreuzer

**Comment by htreu** [comment]
One more thing we need to consider using `DynamicCommandOptionProvider` to support dynamic commands: The handler must then also be able to execute these dynamic commands. This might need some additional work for a plugin mechanism.

**Comment by htreu** [comment]
> I'd tend to saying that if a channel is readOnly=false and has state options, those should be treated as valid command options as well, unless there is a dedicated list of command options provided.

@kaikreuzer what do you think is the benefit of having state options along command options? With state options only we already have the rendering of drop downs and the correlation of state and the selected option.

> Wrt rendering buttons: So far, we render drop-down boxes, which is also a valid solution - it only fails for "pure command channels" which do not have any state and thus a selection should not stay active. But this is going to be addressed with #5011.

I see this PR to be used for "pure command channels" and to me it fits nice into the other channel "types" we already offer:

* trigger channels: the device triggers an event but has no state
* read-only channels: the state is only given by the device
* "normal" channels: state is given by the device but may be set by the framework/user
* "pure command" channels: the opposite of a trigger channel, commands are sent to the device but a state is not received on this channel. Other channels might be affected by the command and update their state accordingly.

**Comment by kaikreuzer** [comment]
> what do you think is the benefit of having state options along command options? 

You mean, if both are identical or in general? In general, they can be different sets. If they are identical, my suggestion was that it should be enough for the user to provide state options and we automatically consider the command options to be the same.

> With state options only we already have the rendering of drop downs

Which I'd consider a bug/workaround of the current UIs. Drop-downs do not make any sense for states, only for commands. So the REST API should provide command options and those should be rendered by the UIs as drop-downs. We should not move the logic to UIs, i.e. they must not have to check whether there are command options and if not, they instead use state options as commands. Much better to have this centrally in the framework.

> I see this PR to be used for "pure command channels" and to me it fits nice into the other channel 

Not sure how your remark is related to my comment that drop-down boxes do not seem to be a good widget for command-only channels...?

**Comment by htreu** [comment]
Thanks @kaikreuzer, that makes it clearer now.

What about different state and command options. What is a UI supposed to render if the current item state has a corresponding state option but no command option? In general, what are the state options then used for? We would need a new set of UI components which make a difference between state display and commands the user can issue. Like: 1st row displays all state options with the current state highlighted, 2nd row shows command buttons with the current state visualised as a pressed button (if any matches).

I will update the PR, so command options will be filled from state options when not explicitly given.

**Comment by kaikreuzer** [comment]
> Like: 1st row displays all state options

There is no need to display state options at all - a "normal" UI could just ignore them, imho they are mostly relevant for rule editors to know what states they might come across and thus offer according triggers etc.

So only the current state is to be shown. And yes, if there is a state that is NOT a command at the same time, the drop-down widget isn't the right choice either. Or, similarly to command-only channels, it could be left with a "blank" option that is displayed and the state simply appears textually like it is actually already possible in e.g. the Basic UI with buttons:

screenshot 2019-01-18 at 11 15 20

**Pull request merged**